lib/proxy/client.go

```diff
@@ -0,0 +1,297 @@
+// Copyright 2022 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"sync"
+	"time"
+
+	clientapi "github.com/gravitational/teleport/api/client/proto"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/metadata"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+const (
+	errorTunnelNotFound = "NOT_FOUND"
+	errorTunnelInvalid  = "INVALID_STATE"
+
+	defaultCleanupInterval = time.Minute * 1
+)
+
+// ClientConfig configures a Client instance.
+type ClientConfig struct {
+	// AccessCache is the caching client connected to the proxy client.
+	AccessCache auth.AccessCache
+	// TLSConfig is the proxy client TLS configuration.
+	TLSConfig *tls.Config
```

Comment:
I am thinking we will need to get the RootCAs here similar to what we are doing on the server side to handle certificate rotation. Based on [this](https://github.com/golang/go/issues/22836) issue it seems like there is no `GetConfigForClient` equivalent on the client side of things. But there are some other ideas to accomplish the same thing in the comments of that issue.

Reply:
Yes I briefly looked into it, that's why I had the access cache added to the config. I'll dig in deeper and see what I can do

Reply:
2df5b1670585c716e35e18fad703f6e12c0cee41